### PR TITLE
Fix rectangle digitizing tools on a raster layer (issue #33583)

### LINKED ISSUE #33583: Crash - using shape digitizing tools on a raster layer
**Describe the bug**

If you use the shape digitizing toolbar (not the plugin) on a raster layer, QGIS crashes.

**How to Reproduce**

1. Load into a new project
   1. a raster layer
   1. a polygon vector layer
1. Select the vector and switch to edit mode
1. Enable the shape digitizing toolbar
1. Select the raster layer
1. Draw a shape with the previously mentionned toolbar
   1. finish your shape with a right click

```
Crash ID: 419df0013f72d85e8071f5ab69345c4623f0fd4c
Stack Trace

QgsPointPatternFillSymbolLayer::hasDataDefinedProperties :
QgsLabelEngineConfigDialog::trUtf8 :
QgsLabelEngineConfigDialog::trUtf8 :
QgsLongLongValidator::QgsLongLongValidator :
QgsLongLongValidator::QgsLongLongValidator :
QWidget::event :
QFrame::event :
QGraphicsView::viewportEvent :
QCoreApplicationPrivate::sendThroughObjectEventFilters :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsPointPatternFillSymbolLayer::hasDataDefinedProperties :
QCoreApplication::notifyInternal2 :
QApplicationPrivate::sendMouseEvent :
QSizePolicy::QSizePolicy :
QSizePolicy::QSizePolicy :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsPointPatternFillSymbolLayer::hasDataDefinedProperties :
QCoreApplication::notifyInternal2 :
QGuiApplicationPrivate::processMouseEvent :
QWindowSystemInterface::sendWindowSystemEvents :
QEventDispatcherWin32::processEvents :
UserCallWinProcCheckWow :
DispatchMessageWorker :
QEventDispatcherWin32::processEvents :
qt_plugin_query_metadata :
QEventLoop::exec :
QCoreApplication::exec :
QgsLabelEngineConfigDialog::trUtf8 :
BaseThreadInitThunk :
RtlUserThreadStart :
```

```
QGIS Info
QGIS Version: 3.11.0-Master
QGIS code revision: 0d84e35839
Compiled against Qt: 5.11.2
Running against Qt: 5.11.2
Compiled against GDAL: 3.1.0dev
Running against GDAL: 3.1.0dev
```

```
System Info
CPU Type: x86_64
Kernel Type: winnt
Kernel Version: 10.0.14393
```

### PR DESCRIPTION
**Pull request by obrix**
Hi,

My name is Bertrand Rix and I'm a new member of Oslandia team. I am mainly a c++ developer. Here is my first contributation to qgis. Any tips is welcomed.

## Description

Fix #33583

      * Avoid casting the current layer to a QgsVectorLayer pointer, this resulted to a null pointer when the current layer isn't one.
      * A similar strategy as the one used in QgsMapToolAddRegularPolygon::deactivate has been used to handle the z value.

This is a simple fix to avoid the app to crash. An evolution could be to completely block the access to the shape digitizing tool when the current layer is not a vector layer.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

File: src/app/qgsmaptooladdrectangle.cpp [hunk 1]
```diff
@@ -71,7 +71,20 @@ void QgsMapToolAddRectangle::deactivate( )
   }
 
   mParentTool->clearCurve( );
-  mParentTool->addCurve( mRectangle.toLineString( !QgsWkbTypes::hasZ( qobject_cast<QgsVectorLayer *>( mCanvas->currentLayer() )->wkbType() ) ) );
+  // keep z value from the first snapped point
+  std::unique_ptr<QgsLineString> ls( mRectangle.toLineString() );
+  for ( const QgsPoint point : qgis::as_const( mPoints ) )
```
Comment (elpaso) [inline src/app/qgsmaptooladdrectangle.cpp]:
```suggestion
  for ( const QgsPoint &point : qgis::as_const( mPoints ) )
```

File: src/app/qgsmaptooladdrectangle.cpp [hunk 2]
```diff
@@ -71,7 +71,20 @@ void QgsMapToolAddRectangle::deactivate( )
   }
 
   mParentTool->clearCurve( );
-  mParentTool->addCurve( mRectangle.toLineString( !QgsWkbTypes::hasZ( qobject_cast<QgsVectorLayer *>( mCanvas->currentLayer() )->wkbType() ) ) );
+  // keep z value from the first snapped point
+  std::unique_ptr<QgsLineString> ls( mRectangle.toLineString() );
```
Comment (elpaso) [inline src/app/qgsmaptooladdrectangle.cpp]: Just for readability, maybe better:
```suggestion
  std::unique_ptr<QgsLineString> lineString( mRectangle.toLineString() );
```

**Comment by elpaso** [comment]
@obrix welcome on board!

**Review by lbartoletti — APPROVED** [review]
LGTM with the suggestions from alessandro. Thank you @obrix

File: src/app/qgsmaptooladdrectangle.cpp (hunk 2 above)
Comment (obrix) [reply src/app/qgsmaptooladdrectangle.cpp]: Changes pushed on the branch fix_33583. Thanks !

File: src/app/qgsmaptooladdrectangle.cpp (hunk 1 above)
Comment (obrix) [reply src/app/qgsmaptooladdrectangle.cpp]: Changes pushed on the branch fix_33583. Thanks !

**Comment by nyalldawson** [comment]
> This is a simple fix to avoid the app to crash. An evolution could be to completely block the access to the shape digitizing tool when the current layer is not a vector layer.

I think this is the real fix for this bug :stuck_out_tongue: . There's already a lot of code in qgisapp.cpp which deals with similar circumstances (See QgisApp::activateDeactivateLayerRelatedActions )

**Comment by obrix** [comment]
As suggested by @nyalldawson I disabled the tools when the current layer is a raster layer. With some notifications to the user when accidentally switching layer while drawing a shape.

File: src/app/qgisapp.cpp [hunk 3]
```diff
@@ -13655,6 +13655,22 @@ void QgisApp::activateDeactivateLayerRelatedActions( QgsMapLayer *layer )
...
+      mActionEllipseCenterPoint->setEnabled( false );
+      mActionEllipseExtent->setEnabled( false );
+      mActionEllipseFoci->setEnabled( false );
+      mActionRectangleCenterPoint->setEnabled( false );
+      mActionRectangleExtent->setEnabled( false );
+      mActionRectangle3PointsDistance->setEnabled( false );
+      mActionRectangle3PointsProjected->setEnabled( false );
+      mActionRegularPolygon2Points->setEnabled( false );
+      mActionRegularPolygonCenterPoint->setEnabled( false );
+      mActionRegularPolygonCenterCorner->setEnabled( false );
```
Comment (lbartoletti) [inline src/app/qgisapp.cpp]: While you're here, please can you add:
```
      mActionReverseLine->setEnabled( false );
      mActionTrimExtendFeature->setEnabled( false );
```

File: src/app/qgsmaptoolcircle2points.cpp [hunk 4]
```diff
@@ -32,6 +32,11 @@ void QgsMapToolCircle2Points::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
 {
   QgsPoint point = mapPoint( *e );
 
+  if ( !currentVectorLayer() )
+  {
+    notifyNotVectorLayer();
```
Comment (lbartoletti) [inline src/app/qgsmaptoolcircle2points.cpp]: Can you clear the rubberband please?
bug webm

**Comment by obrix** [comment]
@elpaso Could you check that the last commits are convenient answers to your comments ?

**Comment by elpaso** [comment]
@obrix LGTM, thanks.

**Comment by mhugo** [comment]
Reviewers seem ok, I'll merge. Thanks @obrix

**Pull request merged**